File: WalkTracker.java
// WalkTracker.java
import java.io.*;
import java.nio.file.*;
import java.time.*;
import java.time.format.*;
import java.util.*;
import com.google.gson.*;

class Config {
    double weight = 70.0;
}

class Walk {
    String date;
    double distance;
    double duration;
    double speed;
    double weight;
    double calories;
}

public class WalkTracker {
    private static final String CONFIG_FILE = "walk_config.json";
    private static final String DATA_FILE = "walks.json";
    private static double weight = 70.0;
    private static List<Walk> walks = new ArrayList<>();
    private static final Gson gson = new GsonBuilder().setPrettyPrinting().create();

    static void loadConfig() {
        try {
            Path path = Paths.get(CONFIG_FILE);
            if (Files.exists(path)) {
                String json = new String(Files.readAllBytes(path));
                Config cfg = gson.fromJson(json, Config.class);
                weight = cfg.weight;
            }
        } catch (Exception e) {}
    }

    static void saveConfig() {
        try {
            Config cfg = new Config();
            cfg.weight = weight;
            Files.write(Paths.get(CONFIG_FILE), gson.toJson(cfg).getBytes());
        } catch (Exception e) {}
    }

    static void loadWalks() {
        try {
            Path path = Paths.get(DATA_FILE);
            if (Files.exists(path)) {
                String json = new String(Files.readAllBytes(path));
                Walk[] arr = gson.fromJson(json, Walk[].class);
                walks = new ArrayList<>(Arrays.asList(arr));
            }
        } catch (Exception e) {}
    }

    static void saveWalks() {
        try {
            Files.write(Paths.get(DATA_FILE), gson.toJson(walks).getBytes());
        } catch (Exception e) {}
    }

    static void addWalk(double dist, double dur, Double w) {
        if (w == null) w = weight;
        double speed = dist / (dur / 60.0);
        double met;
        if (speed < 3.0) met = 2.0;
        else if (speed < 5.0) met = 3.0;
        else if (speed < 6.5) met = 3.8;
        else if (speed < 8.0) met = 5.0;
        else met = 6.0;
        double cal = met * w * (dur / 60.0);
        Walk entry = new Walk();
        entry.date = Instant.now().toString();
        entry.distance = dist;
        entry.duration = dur;
        entry.speed = speed;
        entry.weight = w;
        entry.calories = Math.round(cal * 10) / 10.0;
        walks.add(entry);
        saveWalks();
        System.out.printf("✅ Walk added: %.2f km in %.0f min, %.1f kcal\n", dist, dur, entry.calories);
    }

    static void listWalks() {
        if (walks.isEmpty()) {
            System.out.println("No walks recorded.");
            return;
        }
        System.out.println("\n📋 Walks:");
        for (int i = 0; i < walks.size(); i++) {
            Walk w = walks.get(i);
            String dt = w.date.substring(0, 16).replace('T', ' ');
            System.out.printf("%d. %s | %.2f km | %.0f min | %.1f km/h | %.1f kcal\n",
                i+1, dt, w.distance, w.duration, w.speed, w.calories);
        }
    }

    static void stats() {
        if (walks.isEmpty()) {
            System.out.println("No walks yet.");
            return;
        }
        double totalDist = 0, totalDur = 0, totalCal = 0;
        for (Walk w : walks) {
            totalDist += w.distance;
            totalDur += w.duration;
            totalCal += w.calories;
        }
        System.out.println("\n🚶 Walk Summary");
        System.out.printf("Total walks: %d\n", walks.size());
        System.out.printf("Total distance: %.2f km\n", totalDist);
        System.out.printf("Total duration: %.0f min (%.1f h)\n", totalDur, totalDur/60);
        System.out.printf("Total calories: %.0f kcal\n", totalCal);
    }

    static void exportCSV(String filename) throws IOException {
        Path path = Paths.get(filename);
        try (BufferedWriter writer = Files.newBufferedWriter(path)) {
            writer.write("Date,Distance (km),Duration (min),Speed (km/h),Weight (kg),Calories (kcal)\n");
            for (Walk w : walks) {
                writer.write(String.format("%s,%.2f,%.0f,%.1f,%.1f,%.1f\n",
                    w.date, w.distance, w.duration, w.speed, w.weight, w.calories));
            }
        }
        System.out.printf("Exported %d walks to %s\n", walks.size(), filename);
    }

    public static void main(String[] args) throws Exception {
        loadConfig();
        loadWalks();
        if (args.length < 1) {
            System.out.println("Usage: WalkTracker [weight|add|list|stats|export]");
            return;
        }
        String cmd = args[0];
        switch (cmd) {
            case "weight":
                if (args.length < 2) { System.out.println("Usage: weight <kg>"); return; }
                weight = Double.parseDouble(args[1]);
                saveConfig();
                System.out.printf("Default weight set to %.1f kg\n", weight);
                break;
            case "add":
                if (args.length < 3) { System.out.println("Usage: add <distance> <duration> [--weight <kg>]"); return; }
                double dist = Double.parseDouble(args[1]);
                double dur = Double.parseDouble(args[2]);
                Double w = null;
                for (int i = 3; i < args.length; i++) {
                    if (args[i].equals("--weight") && i+1 < args.length) {
                        w = Double.parseDouble(args[i+1]);
                        i++;
                    }
                }
                addWalk(dist, dur, w);
                break;
            case "list":
                listWalks();
                break;
            case "stats":
                stats();
                break;
            case "export":
                String filename = args.length > 1 ? args[1] : "walks.csv";
                exportCSV(filename);
                break;
            default:
                System.out.println("Unknown command");
        }
    }
}

File: walk_tracker.go
// walk_tracker.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"time"
)

type Walk struct {
	Date     string  `json:"date"`
	Distance float64 `json:"distance"`
	Duration float64 `json:"duration"`
	Speed    float64 `json:"speed"`
	Weight   float64 `json:"weight"`
	Calories float64 `json:"calories"`
}

type Config struct {
	Weight float64 `json:"weight"`
}

var configFile = "walk_config.json"
var dataFile = "walks.json"
var weight = 70.0
var walks []Walk

func loadConfig() {
	data, err := os.ReadFile(configFile)
	if err != nil {
		return
	}
	var cfg Config
	if err := json.Unmarshal(data, &cfg); err == nil {
		weight = cfg.Weight
	}
}

func saveConfig() {
	cfg := Config{Weight: weight}
	data, _ := json.MarshalIndent(cfg, "", "  ")
	os.WriteFile(configFile, data, 0644)
}

func loadWalks() {
	data, err := os.ReadFile(dataFile)
	if err != nil {
		return
	}
	json.Unmarshal(data, &walks)
}

func saveWalks() {
	data, _ := json.MarshalIndent(walks, "", "  ")
	os.WriteFile(dataFile, data, 0644)
}

func addWalk(dist, dur, w float64) {
	if w == 0 {
		w = weight
	}
	speed := dist / (dur / 60.0)
	var met float64
	switch {
	case speed < 3.0:
		met = 2.0
	case speed < 5.0:
		met = 3.0
	case speed < 6.5:
		met = 3.8
	case speed < 8.0:
		met = 5.0
	default:
		met = 6.0
	}
	cal := met * w * (dur / 60.0)
	entry := Walk{
		Date:     time.Now().Format(time.RFC3339),
		Distance: dist,
		Duration: dur,
		Speed:    speed,
		Weight:   w,
		Calories: cal,
	}
	walks = append(walks, entry)
	saveWalks()
	fmt.Printf("✅ Walk added: %.2f km in %.0f min, %.1f kcal\n", dist, dur, cal)
}

func listWalks() {
	if len(walks) == 0 {
		fmt.Println("No walks recorded.")
		return
	}
	fmt.Println("\n📋 Walks:")
	for i, w := range walks {
		dt := w.Date[:16] // approx
		fmt.Printf("%d. %s | %.2f km | %.0f min | %.1f km/h | %.1f kcal\n",
			i+1, dt, w.Distance, w.Duration, w.Speed, w.Calories)
	}
}

func stats() {
	if len(walks) == 0 {
		fmt.Println("No walks yet.")
		return
	}
	var totalDist, totalDur, totalCal float64
	for _, w := range walks {
		totalDist += w.Distance
		totalDur += w.Duration
		totalCal += w.Calories
	}
	fmt.Println("\n🚶 Walk Summary")
	fmt.Printf("Total walks: %d\n", len(walks))
	fmt.Printf("Total distance: %.2f km\n", totalDist)
	fmt.Printf("Total duration: %.0f min (%.1f h)\n", totalDur, totalDur/60)
	fmt.Printf("Total calories: %.0f kcal\n", totalCal)
}

func exportCSV(filename string) {
	f, err := os.Create(filename)
	if err != nil {
		fmt.Println("Error creating file:", err)
		return
	}
	defer f.Close()
	f.WriteString("Date,Distance (km),Duration (min),Speed (km/h),Weight (kg),Calories (kcal)\n")
	for _, w := range walks {
		f.WriteString(fmt.Sprintf("%s,%.2f,%.0f,%.1f,%.1f,%.1f\n",
			w.Date, w.Distance, w.Duration, w.Speed, w.Weight, w.Calories))
	}
	fmt.Printf("Exported %d walks to %s\n", len(walks), filename)
}

func main() {
	loadConfig()
	loadWalks()

	if len(os.Args) < 2 {
		fmt.Println("Usage: walk_tracker [weight|add|list|stats|export]")
		return
	}
	cmd := os.Args[1]

	switch cmd {
	case "weight":
		if len(os.Args) != 3 {
			fmt.Println("Usage: weight <kg>")
			return
		}
		var w float64
		fmt.Sscanf(os.Args[2], "%f", &w)
		weight = w
		saveConfig()
		fmt.Printf("Default weight set to %.1f kg\n", weight)

	case "add":
		if len(os.Args) < 4 {
			fmt.Println("Usage: add <distance> <duration> [--weight <kg>]")
			return
		}
		var dist, dur, w float64
		fmt.Sscanf(os.Args[2], "%f", &dist)
		fmt.Sscanf(os.Args[3], "%f", &dur)
		// Check optional weight
		for i := 4; i < len(os.Args); i++ {
			if os.Args[i] == "--weight" && i+1 < len(os.Args) {
				fmt.Sscanf(os.Args[i+1], "%f", &w)
				break
			}
		}
		addWalk(dist, dur, w)

	case "list":
		listWalks()

	case "stats":
		stats()

	case "export":
		filename := "walks.csv"
		if len(os.Args) >= 3 {
			filename = os.Args[2]
		}
		exportCSV(filename)

	default:
		fmt.Println("Unknown command")
	}
}
